### PR TITLE
wazevo: add nontrapping fp conversions (wasm v2)

### PR DESCRIPTION
**Pull request by evacchi**
moves forward #1496 with support for non-trapping floating point conversions to integer.

This also changes the signature of `lowerOpcode(op)` to `lowerCurrentOpcode()`. In order to parse `wasm.OpcodeMisc*TruncSat*` etc. it is necessary to parse `wasm.OpcodeMiscPrefix` first, and then advance to the next opcode, which will be in the range `[0x00..0x07]` They cannot be parsed otherwise, because then these opcodes would be mismatched in the `switch` statement for other unrelated opcodes.

Now `lowerCurrentOpcode()` instead moves `pc`itself; for this reason, all the related logic such as the debugging statements are now moved to this procedure, and  `return` statements had had to be changed to `break`s, otherwise the final `pc++` would be skipped in many cases.

### CODE REVIEW

File: internal/engine/wazevo/backend/backend_test.go [hunk 1]
```diff
@@ -1713,6 +1713,63 @@ L1 (SSA Block: blk0):
 	fcvt d1, s1
 	ldr x30, [sp], #0x10
 	ret
+`,
+		},
+		{
+			name: "nontrapping_float_conversions",
+			m:    testcases.NonTrappingFloatConversions.Module,
+			afterLoweringARM64: `
```
Comment (mathetake) [inline internal/engine/wazevo/backend/backend_test.go:1721]: nothing interesting here, so let's not have `afterLoweringARM64` and only have `afterFinalizeARM64` (the assertion on `afterLoweringARM64` will be skipped if it is empty).

File: internal/engine/wazevo/frontend/lower.go
Comment (mathetake) [inline internal/engine/wazevo/frontend/lower.go:1]: could you revert this change?

Comment (evacchi) [reply internal/engine/wazevo/frontend/lower.go:1]: this was intended, `lowerOpcode(op)` operates on a given opcode, but in order to parse `wasm.OpcodeMisc*TruncSat*` etc. it is necessary to parse `wasm.OpcodeMiscPrefix` and then advance to the next opcode. Otherwise the opcodes for `wasm.OpcodeMisc*TruncSat*` clash with other unrelated opcodes. 

So I moved the logic to advance `pc` inside `lowerOpcode(op)`, and renamed it to `lowerCurrentOpcode()`. I also moved all the related logic such as the debugging statements.

I can do it differently, this seemed the most straightforward way.

Comment (mathetake) [reply internal/engine/wazevo/frontend/lower.go:1]: ah ok make sense! thank you for clarifying

Comment (evacchi) [reply internal/engine/wazevo/frontend/lower.go:1]: adding this to the top post for future reference

File: internal/engine/wazevo/frontend/lower.go [hunk 2]
```diff
@@ -475,18 +470,49 @@ func (c *Compiler) lowerOpcode(op wasm.Opcode) {
...
 		).Insert(builder).Return()
 		state.push(ret)
+	case wasm.OpcodeMiscPrefix:
+		c.loweringState.pc++
+		// A misc opcode is encoded as an unsigned variable 32-bit integer.
+		miscOpUint, num, err := leb128.LoadUint32(c.wasmFunctionBody[c.loweringState.pc:])
+		if err != nil {
+			// fixme return err
+			panic(fmt.Sprintf("failed to read misc opcode: %v", err))
+		}
```
Comment (evacchi) [inline internal/engine/wazevo/frontend/lower.go:490]: in current compiler the procedure returns an error, should we do it here too? @mathetake

Comment (mathetake) [reply internal/engine/wazevo/frontend/lower.go:490]: no, this comes after function validation, so panic should be fine and shouldn't happen. See the code pass of wazero.Runtime.CompileModule!

File: internal/engine/wazevo/backend/backend_test.go (hunk 1 above)
Comment (mathetake) [reply internal/engine/wazevo/backend/backend_test.go:1730]: @evacchi I believe this hasn't be resolved.

Comment (evacchi) [reply internal/engine/wazevo/backend/backend_test.go:1730]: maybe I reset it by mistake

File: internal/engine/wazevo/backend/isa/arm64/lower_instr.go [hunk 3]
```diff
@@ -427,7 +429,7 @@ func (m *machine) lowerFpuUniOp(op fpuUniOp, in, out ssa.Value) {
 	m.insert(neg)
 }
 
-func (m *machine) lowerFpuToInt(rd, rn operand, ctx regalloc.VReg, signed, src64bit, dst64bit bool) {
+func (m *machine) lowerFpuToInt(rd, rn operand, ctx regalloc.VReg, signed, src64bit, dst64bit, nonTrapping bool) {
 	// First of all, we have to clear the FPU flags.
```
Comment (mathetake) [inline internal/engine/wazevo/backend/isa/arm64/lower_instr.go:433]: I believe this shouldn't be neede when non trapping

Comment (evacchi) [reply internal/engine/wazevo/backend/isa/arm64/lower_instr.go:433]: that's ok, I only copied it because the original compiler is doing it

https://github.com/tetratelabs/wazero/blob/ea634990050bbe9af5ce230cce9811d54566c60f/internal/engine/compiler/impl_arm64.go#L2189-L2192

at this point I'd rather create a `lowerFpuToIntNontrapping()` since it becomes three lines. WDYT? or maybe add `lowerFpuToIntTrapping()` which invokes `lowerFpuToInt()` that becomes the 3-liner:


```
func lowerFpuToIntTrapping(...) {
   // clear flags
   lowerFpuToInt(...)
   // handle trap
}
func lowerFpuToInt(...) {
	cvt := m.allocateInstr()
	cvt.asFpuToInt(rd, rn, signed, src64bit, dst64bit)
	m.insert(cvt)
}
```

Comment (mathetake) [reply internal/engine/wazevo/backend/isa/arm64/lower_instr.go:433]: up to you!

Comment (mathetake) [reply internal/engine/wazevo/backend/isa/arm64/lower_instr.go:433]: either works for me 🍷

Comment (evacchi) [reply internal/engine/wazevo/backend/isa/arm64/lower_instr.go]: I am keeping it simple (just if statements) because I guess I am still too jet-lagged to refactor it right :D basically the ctx register becomes useless for the nontrapping version, but that requires a few more changes to bubble up the chain and something was going wrong at the regalloc stage. I can refactor it later.

Comment (evacchi) [reply internal/engine/wazevo/backend/isa/arm64/lower_instr.go]: I am marking this as ready, but if you'd rather me do the refactor I can do push it here later

**Review by mathetake — APPROVED** [review]
💯
